src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/CodeRenderingContext.cs

```diff
@@ -10,114 +11,59 @@ public abstract class CodeRenderingContext
         internal static readonly object NewLineString = "NewLineString";
         internal static readonly object SuppressUniqueIds = "SuppressUniqueIds";
 
-        public static CodeRenderingContext Create(RazorCodeDocument codeDocument, RazorCodeGenerationOptions options)
-        {
-            if (codeDocument == null)
-            {
-                throw new ArgumentNullException(nameof(codeDocument));
-            }
-
-            if (options == null)
-            {
-                throw new ArgumentNullException(nameof(options));
-            }
-
-            IntermediateNodeWriter nodeWriter;
-            TagHelperWriter tagHelperWriter;
-
-            if (options.DesignTime)
-            {
-                nodeWriter = new DesignTimeNodeWriter();
-                tagHelperWriter = new DesignTimeTagHelperWriter();
-            }
-            else
-            {
-                nodeWriter = new RuntimeNodeWriter();
-                tagHelperWriter = new RuntimeTagHelperWriter();
-            }
-
-            var documentKind = codeDocument.GetDocumentIntermediateNode()?.DocumentKind;
-            var codeWriter = new CodeWriter();
-            var context = new DefaultCodeRenderingContext(codeWriter, nodeWriter, documentKind, codeDocument.Source, options)
-            {
-                TagHelperWriter = tagHelperWriter
-            };
+        public abstract CodeWriter CodeWriter { get; }
 
-            var newLineString = codeDocument.Items[NewLineString];
-            if (newLineString != null)
-            {
-                // Set new line character to a specific string regardless of platform, for testing purposes.
-                codeWriter.NewLine = (string)newLineString;
-            }
+        public abstract RazorDiagnosticCollection Diagnostics { get; }
 
-            context.Items[SuppressUniqueIds] = codeDocument.Items[SuppressUniqueIds];
+        public abstract string DocumentKind { get; }
 
-            return context;
-        }
+        public abstract ItemCollection Items { get; }
 
-        public abstract CodeWriter CodeWriter { get; }
+        public abstract IntermediateNodeWriter NodeWriter { get; }
 
-        public abstract IntermediateNodeWriter NodeWriter { get; protected set; }
+        public abstract RazorCodeGenerationOptions Options { get; }
 
         public abstract RazorSourceDocument SourceDocument { get; }
 
-        public abstract RazorCodeGenerationOptions Options { get; }
+        public abstract Scope CreateScope();
 
-        public abstract RazorDiagnosticCollection Diagnostics { get; }
+        public abstract Scope CreateScope(IntermediateNodeWriter writer);
```

Comment:
Note for others looking at this CreateScope/EndScope will go away in follow up prs.